### PR TITLE
[TensorParallel] Add ParallelStyle base class and input preparation APIs

### PR DESCRIPTION
**Pull request by XilunWu**
Goal:

- This PR serves as a starting point of adding Junjie's TensorParallelism low-level and high-level APIs for enabling composable 2D parallelism with DTensor.

Items:

- added ParallelStyle base class
- added input preparation APIs `shard_input_1d` and `replicate_input_1d`.
- added unit tests for `shard_input_1d` and `replicate_input_1d`.

### CODE REVIEW

File: spmd/tensor/parallel/api.py
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:92]: For output, we might also need a DeviceMesh, let's make it Optional. Also, since we need to pass in the dimension, can we also updated the signature here? I have updated the design already.

File: spmd/tensor/parallel/api.py
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:83]: I feel we can move this into another file named parallel_style.py? WDY?

File: spmd/tensor/parallel/api.py
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:17]: We need to add comments for this API?

File: spmd/tensor/parallel/api.py
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:127]: This is still needed for 2D. We should still keep it.

**Review by fduwjj — COMMENTED** [review]
Please add test for each API we added here. Thanks!

File: spmd/tensor/parallel/api.py
Comment (XilunWu) [reply spmd/tensor/parallel/api.py:83]: it makes sense. Also put `MakeInputShard` etc. in parallel_style.py` maybe?

File: spmd/tensor/parallel/api.py
Comment (wanchaol) [inline spmd/tensor/parallel/api.py:129]: This is wrong, can you check how `isinstance` should be used?

File: spmd/tensor/parallel/api.py
Comment (wanchaol) [inline spmd/tensor/parallel/api.py:131]: we shouldn't disclose the function information in error message as we don't expose it to the user.  it should just be expecting torch.nn.Linear but found ...

File: spmd/tensor/parallel/api.py
Comment (wanchaol) [inline spmd/tensor/parallel/api.py:136]: same comment as above.

File: spmd/tensor/parallel/api.py
Comment (wanchaol) [inline spmd/tensor/parallel/api.py:146]: all of `isinstance` check are wrong. Could you please add some tests for this API?

**Review by wanchaol — CHANGES_REQUESTED** [review]
Please add some tests to ensure the API works

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:81]: For better debuggabilty, can we say something like "Input to tensor parallel module is {} not torch.Tensor nor DTensor!".format(type(input)) What do you think?

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:61]: Can we call it input? Up to you.

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:72]: ditto.

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:14]: Ideally you can split this one (abstract class) into a separate PR. But this is fine this time. 

Also can we extend ABC instead? (https://docs.python.org/3/library/abc.html)

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:16]: Nit: to be parallelized.

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:56]: See my comment in L81.

File: test/spmd/tensor/parallel/test_tp_examples.py
Comment (fduwjj) [inline test/spmd/tensor/parallel/test_tp_examples.py:125]: Nit: you can separate error case and normal cases into two tests. But it's up to you.

File: test/spmd/tensor/parallel/test_tp_examples.py
Comment (fduwjj) [inline test/spmd/tensor/parallel/test_tp_examples.py:166]: Are you sure this condition will succeed? The size of tensor is [8, 16] while local tensor size has size of [2, 16]. Or am I missing something?

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:37]: Now I see the problem. Making dim parameter Optional does not make the param len variable. Now I see the disadvantage of this design. Let's remove Optional[int] and pass it in ParalellStyle. What do you think?

**Review by fduwjj — COMMENTED** [review]
Thanks for working on this and leave some comments. LMK if you want to chat more on this. Thanks!

**Comment by fduwjj** [comment]
Also please rebase. Thanks!

File: spmd/tensor/parallel/__init__.py
Comment (fduwjj) [inline spmd/tensor/parallel/__init__.py:12]: Let's use from ... import ... so that we only expose certain APIs. Thansk!

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:34]: Also per PyTorch coding standard, you need to add comment to this function with the spec of input/output.

Comment (fduwjj) [reply spmd/tensor/parallel/style.py:34]: One example for this: https://github.com/pytorch/pytorch/blob/e2f0648750f2d0d0ac648728ce4c514db178cfa1/torch/distributed/_shard/sharded_tensor/api.py#L785.

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:53]: Please use run_check=False for now. Thanks!

File: test/spmd/tensor/parallel/test_tp_examples.py
Comment (fduwjj) [inline test/spmd/tensor/parallel/test_tp_examples.py:125]: Also, can we move the test to a separate file? like test_tp_style.py?

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [reply spmd/tensor/parallel/style.py:60]: After a second thought, I think we can keep it for now since we now only use 1D DeviceMesh for TP. Our design is generic enough for future extension.

Comment (XilunWu) [reply spmd/tensor/parallel/style.py:60]: Agree. When there're use cases of DeviceMesh more than 1D, we can add corresponding shard/replicate functions.

File: spmd/tensor/parallel/style.py
Comment (XilunWu) [reply spmd/tensor/parallel/style.py:14]: TIL Python's Abstract Base Classes.

File: spmd/tensor/parallel/style.py
Comment (XilunWu) [reply spmd/tensor/parallel/style.py:53]: Is it only for saving comms?

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:34]: Also can we make the name: make_input_shard_1d? The reason is that, for _prepare_output, there is no verb for converting to tensor. Let's all use the verb make to unify the component. Thanks!

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:22]: Upon checking, this is fine.

Optional[X] is equivalent to X | None (or Union[X, None]). So let's keep Optional[int].

https://docs.python.org/3/library/typing.html

File: spmd/tensor/parallel/style.py
Comment (XilunWu) [reply spmd/tensor/parallel/style.py:64]: The reason why I put the assertion below is to make sure that `dtensor.device_mesh` won't be None. But this seems not likely I guess. Then I can simply raise RuntimeError above.

Comment (wanchaol) [reply spmd/tensor/parallel/style.py:64]: DTensor can't construct without a device_mesh, so we don't need to worry about this :)

**Comment by fduwjj** [comment]
Let's not merge for now and please first focus on your own part. https://github.com/pytorch/tau/pull/615 is not overlapping with your change and there should not be too many merge conflict. Thanks!

File: test/spmd/tensor/parallel/test_tp_style.py
Comment (fduwjj) [inline test/spmd/tensor/parallel/test_tp_style.py:152]: Thanks for adding this.

File: spmd/tensor/parallel/style.py
Comment (XilunWu) [reply spmd/tensor/parallel/style.py:22]: Since `ParallelStyle` is a public API, I don't think we should have type alias (at least a private type alias) in it. What do you think?

**Review by fduwjj — APPROVED** [review]
To unblock, but keep the records of three things:
1. Type of Alias.
2. Make abstraction logic into a decorator.
3. Add more test cases if you think are necessary.

For every item if you don't think you can complete in this PR, please create a task and put your name here. Let's document everything. Thanks!

File: spmd/tensor/parallel/style.py
Comment (fduwjj) [reply spmd/tensor/parallel/style.py:22]: The API itself is a private API and why do we think type alias is not working here? If you want the func signature to be user facing, you can make the type public right?

**Pull request merged**
